src/chain/split.rs

```diff
@@ -0,0 +1,939 @@
+        self,
+        builder: &'b mut Builder<'w, 's, 'a>,
+    ) -> SplitBuilder<'w, 's, 'a, 'b, T> {
+        assert_eq!(self.scope, builder.scope);
+        SplitBuilder {
+            outputs: self,
+            builder,
+        }
+    }
+
+    pub(crate) fn new(scope: Entity, source: Entity) -> Self {
+        Self {
+            scope,
+            source,
+            last_key: None,
+            used: Default::default(),
+        }
+    }
+}
+
+/// This is a type alias for the result returned by chainable [`SplitBuilder`]
+/// functions. If the last connection succeeded, you will receive [`Ok`] with
+/// the [`SplitBuilder`] which you can keep building off of. Otherwise if the
+/// last connection failed, you will receive an [`Err`] with the [`SplitBuilder`]
+/// bundled with [`SplitConnectionError`] to tell you what went wrong. You can
+/// continue building with the [`SplitBuilder`] even if an error occurred.
+///
+/// You can use `ignore_result` from the [`IgnoreSplitChainResult`] trait to
+/// just keep chaining without checking whether the connection suceeded.
+pub type SplitChainResult<'w, 's, 'a, 'b, T> = Result<
+    SplitBuilder<'w, 's, 'a, 'b, T>,
+    (SplitBuilder<'w, 's, 'a, 'b, T>, SplitConnectionError),
+>;
+
+/// A helper trait that allows users to ignore any failures while chaining
+/// connections to a split.
+pub trait IgnoreSplitChainResult<'w, 's, 'a, 'b, T: Splittable> {
+    /// Ignore whether the result of the chaining connection was [`Ok`] or [`Err`]
+    /// and just keep chaining.
+    fn ignore_result(self) -> SplitBuilder<'w, 's, 'a, 'b, T>;
+}
+
+impl<'w, 's, 'a, 'b, T: Splittable> IgnoreSplitChainResult<'w, 's, 'a, 'b, T>
+    for SplitChainResult<'w, 's, 'a, 'b, T>
+{
+    fn ignore_result(self) -> SplitBuilder<'w, 's, 'a, 'b, T> {
+        match self {
+            Ok(split) => split,
+            Err((split, _)) => split,
+        }
+    }
+}
+
+/// Information about why a connection to a split failed
+#[derive(ThisError, Debug, Clone)]
+#[error("An error occurred while trying to connect to a split")]
+pub enum SplitConnectionError {
+    /// The requested index was already connected to.
+    KeyAlreadyUsed,
+    /// The requested index is out of bounds.
+    KeyOutOfBounds,
+}
+
+/// Used by implementers of the [`Splittable`] trait to help them send their
+/// split values to the proper input slots.
+pub struct SplitDispatcher<'a, Key, Item> {
+    pub(crate) connections: &'a HashMap<Key, usize>,
+    pub(crate) outputs: &'a mut Vec<Vec<Item>>,
+}
+
+impl<'a, Key, Item> SplitDispatcher<'a, Key, Item>
+where
+    Key: 'static + Send + Sync + Eq + Hash + Clone + Debug,
+    Item: 'static + Send + Sync,
+{
+    /// Send a value for a key. The split operation will make sure that this
+    /// value will be sent to the target associated with the key.
+    ///
+    /// If there is no connection associated with the specified key, the value
+    /// will be returned as [`Err`].
```

Comment:
I believe it now returns `None`

Reply:
In fact the entire description of the function is out of date. Fixed here: https://github.com/open-rmf/bevy_impulse/pull/33/commits/669027362ac953144ddb5c11c755118aa69a1750